skimage/morphology/footprints.py

```diff
@@ -18,22 +97,68 @@ def square(width, dtype=np.uint8):
 
     Other Parameters
     ----------------
-    dtype : data-type
+    dtype : data-type, optional
         The data type of the footprint.
+    decomposition : {None, 'separable', 'sequence'}, optional
+        If None, a single array is returned. For 'sequence', a tuple of smaller
+        footprints is returned. Applying this series of smaller footprints will
+        given an identical result to a single, larger footprint, but often with
+        better computational performance. See Notes for more details.
+        With 'separable', this function uses separable 1D footprints for each
+        axis. Whether 'seqeunce' or 'separable' is computationally faster may
+        be architecture-dependent.
 
     Returns
     -------
-    footprint : ndarray
-        A footprint consisting only of ones, i.e. every pixel belongs to the
-        neighborhood.
+    footprint : ndarray or tuple
+        The footprint where elements of the neighborhood are 1 and 0 otherwise.
+        When `decomposition` is None, this is just a numpy.ndarray. Otherwise,
+        this will be a tuple whose length is equal to the number of unique
+        structuring elements to apply (see Notes for more detail)
+
+    Notes
+    -----
+    When `decomposition` is not None, each element of the `footprint`
+    tuple is a 2-tuple of the form ``(ndarray, num_iter)`` that specifies a
+    footprint array and the number of iterations it is to be applied.
+
+    For binary morphology, using ``decomposition='sequence'``
+    was observed to give better performance, with the magnitude of the
+    performance increase rapidly increasing with footprint size. For grayscale
+    morphology with square footprints, it is recommended to use
+    ``decomposition=None`` since the internal SciPy functions that are called
+    already have a fast implementation based on separable 1D sliding windows.
+    """
+    if decomposition is None:
+        return np.ones((width, width), dtype=dtype)
+
+    if decomposition == 'separable' or width % 2 == 0:
+        sequence = [(np.ones((width, 1), dtype=dtype), 1),
+                    (np.ones((1, width), dtype=dtype), 1)]
+    elif decomposition == 'sequence':
+        # only handles odd widths
```

Comment:
This behaviour is not documented :confused:

Reply:
I will update this in the docstring. There was also past discussion in #4536 about making all footprints specified via a radius (or half-width) parameter. If we went with that approach, then all of these would always only support odd lengths of `2*radius + 1` (If we do go that route, it should be through a separate PR, rather than adding it onto this one)

Reply:
The docstring has now been updated

Reply:
>  There was also past discussion in #4536 about making all footprints specified via a radius (or half-width) parameter ...

I remember this discussion, but I think that the conclusion was to use the `size` argument name and to introduce the `origin` argument name to handle even values...